hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/RegionHDFSBlockLocationFinder.java

```diff
@@ -102,8 +108,28 @@ public HDFSBlocksDistribution load(RegionInfo key) throws Exception {
    * @return A new Cache.
    */
   private LoadingCache<RegionInfo, HDFSBlocksDistribution> createCache() {
-    return CacheBuilder.newBuilder().expireAfterWrite(CACHE_TIME, TimeUnit.MILLISECONDS)
-      .build(loader);
+
+    long maximumWeight = getConf().getLong(
+      BALANCER_REGIONBLOCKLOCATION_CACHE_MAX_SIZE, 0);
+
+    if (maximumWeight > 0) {
+      return CacheBuilder.newBuilder()
+        .expireAfterWrite(CACHE_TIME, TimeUnit.MILLISECONDS)
+        .weigher(
+          new Weigher<RegionInfo, HDFSBlocksDistribution>() {
+            @Override
+            public int weigh(RegionInfo regionInfo, HDFSBlocksDistribution hdfsBlocksDistribution) {
+              return (int) (regionInfo.heapSize() +
+                hdfsBlocksDistribution.heapSize());
```

Comment:
Calculating the heap involves too many changes, shall we limit the cache with maximumSize?